website/docs/faqs/Project/udfs-vs-macros.md

```diff
@@ -0,0 +1,86 @@
+---
+title: When should I use a UDF instead of a macro?
+description: "Guidance on choosing between UDFs and macros"
+sidebar_label: 'UDFs vs macros'
+id: udfs-vs-macros
+---
+
+Both user-defined functions (UDFs) and macros let you reuse logic across your dbt project, but they work in fundamentally different ways. Here's when to use each:
+
+## Use UDFs when:
+
+<Expandable alt_header="You need logic accessible outside dbt">
+
+UDFs are created in your warehouse and can be used by BI tools, data science notebooks, SQL clients, or any other tool that connects to your warehouse. Macros only work within dbt.
+
+</Expandable>
+
+<Expandable alt_header="You want better performance for complex operations">
+
+UDFs are compiled and optimized by your warehouse's query engine, which can provide better performance for compute-intensive operations compared to Jinja macros that generate SQL text.
+
+</Expandable>
+
+<Expandable alt_header="You need to standardize warehouse-native functions">
+
+UDFs let you create reusable warehouse functions for data validation, custom formatting, or business-specific calculations that need to be consistent across all your data tools.
+
+</Expandable>
+
+<Expandable alt_header="Your logic is SQL-based and doesn't need Jinja templating">
+
+If your reusable logic is pure SQL without needing dynamic SQL generation, a UDF is cleaner and more portable.
+
+</Expandable>
+
+## Use macros when:
+
+<Expandable alt_header="You need to generate SQL dynamically">
+
+Macros excel at generating SQL based on conditions, looping through lists, or building queries programmatically. UDFs can't do this.
+
+</Expandable>
+
+<Expandable alt_header="You want to generate DDL or DML statements">
+
+Macros can create entire model definitions, tests, or any SQL statement. UDFs are limited to returning values or tables.
+
+</Expandable>
+
+<Expandable alt_header="You need to adapt SQL across different warehouses">
+
+Macros can use Jinja logic to generate warehouse-specific SQL, making your dbt project portable across platforms. UDFs are warehouse-specific.
```

Comment:
```suggestion
Macros can use Jinja conditional logic to generate SQL that's dependent on which warehouse you're using (see [cross-database macros](https://docs.getdbt.com/reference/dbt-jinja-functions/cross-database-macros)), making your dbt project portable across platforms. UDFs are warehouse-specific.
```

Reply:
@runleonarun i don't think this is the right way to do linking, but i attempted!

Reply:
actually - i don't think this is true - UDFs can also use jinja!

cc: @QMalcolm

Reply:
though for UDFs you do have to provide warehouse-specific argument `data_type`s